library/std/src/sys/common/thread_local/fast_local.rs

```diff
@@ -1,70 +1,91 @@
     // used to generate the `LocalKey` value for const-initialized thread locals
     (@key $t:ty, const $init:expr) => {{
+        const __INIT: $t = $init;
+
         #[inline]
         #[deny(unsafe_op_in_unsafe_fn)]
         unsafe fn __getit(
             _init: $crate::option::Option<&mut $crate::option::Option<$t>>,
         ) -> $crate::option::Option<&'static $t> {
-            const INIT_EXPR: $t = $init;
-            // If the platform has support for `#[thread_local]`, use it.
-            #[thread_local]
-            static mut VAL: $t = INIT_EXPR;
-
-            // If a dtor isn't needed we can do something "very raw" and
-            // just get going.
-            if !$crate::mem::needs_drop::<$t>() {
-                unsafe {
-                    return $crate::option::Option::Some(&VAL)
-                }
-            }
-
-            // 0 == dtor not registered
-            // 1 == dtor registered, dtor not run
-            // 2 == dtor registered and is running or has run
-            #[thread_local]
-            static STATE: $crate::cell::Cell<$crate::primitive::u8> = $crate::cell::Cell::new(0);
-
-            // Safety: Performs `drop_in_place(ptr as *mut $t)`, and requires
-            // all that comes with it.
-            unsafe extern "C" fn destroy(ptr: *mut $crate::primitive::u8) {
-                $crate::thread::local_impl::abort_on_dtor_unwind(|| {
-                    let old_state = STATE.replace(2);
-                    $crate::debug_assert_eq!(old_state, 1);
-                    // Safety: safety requirement is passed on to caller.
-                    unsafe { $crate::ptr::drop_in_place(ptr.cast::<$t>()); }
-                });
-            }
+            use $crate::thread::local_impl::Storage;
+            use $crate::mem::needs_drop;
+            use $crate::convert::identity;
 
-            unsafe {
-                match STATE.get() {
-                    // 0 == we haven't registered a destructor, so do
-                    //   so now.
-                    0 => {
-                        $crate::thread::local_impl::Key::<$t>::register_dtor(
-                            $crate::ptr::addr_of_mut!(VAL) as *mut $crate::primitive::u8,
-                            destroy,
-                        );
-                        STATE.set(1);
-                        $crate::option::Option::Some(&VAL)
-                    }
-                    // 1 == the destructor is registered and the value
-                    //   is valid, so return the pointer.
-                    1 => $crate::option::Option::Some(&VAL),
-                    // otherwise the destructor has already run, so we
-                    // can't give access.
-                    _ => $crate::option::Option::None,
+            if needs_drop::<$t>() {
+                #[thread_local]
+                static VAL: Storage<$t, $t, ()> = Storage::new(__INIT);
+                unsafe {
+                    VAL.get_or_init(identity)
```

Comment:
```suggestion
                #[thread_local]
                static VAL: Storage<(), $t, ()> = Storage::new(());
                unsafe {
                    VAL.get_or_init(|| __INIT)
```

This enables layout optimizations if the type `$t` has at least two niches, and makes the speculative read in `initialize` more robust.

In fact, with this change  `Storage` always has a unit sized `Initialized`  variant, so we could just make that a unit variant, and `State::new` could take no args. Simplifying the code a bit.

Reply:
In that case the initializing closure in the `const` case would have to copy memory, whereas currently it just has to update the discriminant, because the initial value was already written by the system during TLS construction.